packages/svelte2tsx/src/nodes/ImplicitTopLevelNames.ts

```diff
@@ -0,0 +1,79 @@
+import ts from 'typescript';
+import MagicString from 'magic-string';
+import { getBinaryAssignmentExpr, extractIdentifiers } from '../utils/tsAst';
+
+export class ImplicitTopLevelNames {
+    private map = new Set<ts.LabeledStatement>();
+
+    add(
+        binaryExpr: ts.BinaryExpression,
+        node: ts.LabeledStatement,
+        rootVariables: Set<string>,
+        astOffset: number,
+        str: MagicString,
+    ) {
+        this.map.add(node);
+
+        if (
+            ts.isExpressionStatement(node.statement) &&
+            ts.isParenthesizedExpression(node.statement.expression) &&
+            (ts.isObjectLiteralExpression(binaryExpr.left) ||
+                ts.isArrayLiteralExpression(binaryExpr.left)) &&
+            this.containsNoTopLevelNames(binaryExpr.left, rootVariables)
```

Comment:
I think this part should execute later. At this point, not all the top-level names are found. if the variable is declared later in the code than the $: statement the parenthesize will be removed but the $; label won't be replaced by `let`.

Reply:
True, we should either do it all right here or do it all in the end.

I think we have to do it right here; as stated in the comment afterwards: Because we use `remove` which would remove everything that we appended/prepended previously in that range. This includes things like the transformation of stores.
In TS it will also throw an error if you use the variable before declaring it. I doubt many people do it, and if so, they just need to move the declaration up.

Reply:
I found a way to do the transformation without removing the previously added strings by doing `str.overwrite(start, start + 1, '', { contentOnly: true })` instead of `str.remove(start, start + 1)`